Sources/PostgresNIO/Pool/ConnectionFactory.swift

```diff
@@ -0,0 +1,206 @@
+import Logging
+import NIOConcurrencyHelpers
+import NIOCore
+import NIOSSL
+
+@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
+final class ConnectionFactory: Sendable {
+
+    struct ConfigCache: Sendable {
+        var config: PostgresClient.Configuration
+    }
+
+    let configBox: NIOLockedValueBox<ConfigCache>
+
+    struct SSLContextCache: Sendable {
+        enum State {
+            case none
+            case producing(TLSConfiguration, [CheckedContinuation<NIOSSLContext, any Error>])
+            case cached(TLSConfiguration, NIOSSLContext)
+            case failed(TLSConfiguration, any Error)
+        }
+
+        var state: State = .none
+    }
+
+    let sslContextBox = NIOLockedValueBox(SSLContextCache())
+
+    let eventLoopGroup: any EventLoopGroup
+
+    let logger: Logger
+
+    init(config: PostgresClient.Configuration, eventLoopGroup: any EventLoopGroup, logger: Logger) {
+        self.eventLoopGroup = eventLoopGroup
+        self.configBox = NIOLockedValueBox(ConfigCache(config: config))
+        self.logger = logger
+    }
+
+    func makeConnection(_ connectionID: PostgresConnection.ID, pool: PostgresClient.Pool) async throws -> PostgresConnection {
+        let config = try await self.makeConnectionConfig()
+
+        var connectionLogger = self.logger
+        connectionLogger[postgresMetadataKey: .connectionID] = "\(connectionID)"
+
+        return try await PostgresConnection.connect(
+            on: self.eventLoopGroup.any(),
+            configuration: config,
+            id: connectionID,
+            logger: connectionLogger
+        ).get()
+    }
+
+    func makeConnectionConfig() async throws -> PostgresConnection.Configuration {
+        let config = self.configBox.withLockedValue { $0.config }
+
+        let tls: PostgresConnection.Configuration.TLS
+        switch config.tls.base {
+        case .prefer(let tlsConfiguration):
+            let sslContext = try await self.getSSLContext(for: tlsConfiguration)
+            tls = .prefer(sslContext)
+
+        case .require(let tlsConfiguration):
+            let sslContext = try await self.getSSLContext(for: tlsConfiguration)
+            tls = .require(sslContext)
+        case .disable:
+            tls = .disable
+        }
+
+        var connectionConfig: PostgresConnection.Configuration
+        switch config.endpointInfo {
+        case .bindUnixDomainSocket(let path):
```

Comment:
Come to think of it, how about the "existing `Channel`" thing too?

Reply:
How do we want to use a single Channel for multiple connections. We will need to come up with a custom factory. That is a valuable goal. But not today.